go/database/vt/state_test.go

```diff
@@ -0,0 +1,417 @@
+	testCodes := func(size int) {
+		code := make([]byte, size)
+		for i := 0; i < len(code); i++ {
+			code[i] = byte(i)
+		}
+
+		const numOfCodes = 1000
+		expectedCodes := [numOfCodes][]byte{}
+		update := common.Update{}
+		for i := 0; i < numOfCodes; i++ {
+			addr := common.Address{byte(i), byte(i >> 8)}
+			update.Codes = append(update.Codes, common.CodeUpdate{
+				Account: addr,
+				Code:    code,
+			})
+			expectedCodes[i] = code
+			code = append(code, byte(i))
+		}
+
+		require.NoError(t, state.Apply(0, update), "failed to apply update")
+
+		// check hash consistency
+		hash, err := state.GetHash()
+		require.NoError(t, err, "failed to get hash for block")
+		require.NotEmpty(t, hash, "hash should not be empty for block")
+
+		for i := 0; i < numOfCodes; i++ {
+			addr := common.Address{byte(i), byte(i >> 8)}
+
+			code, err := state.GetCode(addr)
+			require.NoError(t, err, "failed to get code for account %x", addr)
+			require.True(t, bytes.Equal(code, expectedCodes[i]), "unexpected code for account %x: got %v, want %v", addr, code, expectedCodes[i])
+
+			codeHash, err := state.GetCodeHash(addr)
+			require.NoError(t, err, "failed to get code hash for account %x", addr)
+			require.Equal(t, common.Keccak256(expectedCodes[i]), codeHash, "unexpected code hash for account %x", addr)
+
+			codeLen, err := state.GetCodeSize(addr)
+			require.NoError(t, err, "failed to get code size for account %x", addr)
+			require.Equal(t, len(expectedCodes[i]), codeLen, "unexpected code size for account %x", addr)
+		}
+	}
+
+	testCodes(4096)
+	testCodes(2048) // smaller sets overrides larger ones
+	testCodes(5120) // larger sets overrides smaller ones
+}
+
+func TestState_Storage_Leading_Zeros_HashesMatch(t *testing.T) {
+	addr1 := common.Address{0, 0, 0, 1}
+	value := common.Value{0, 0, 0, 1}
+
+	update := common.Update{
+		CreatedAccounts: []common.Address{addr1},
+		Balances: []common.BalanceUpdate{
+			{Account: addr1, Balance: amount.New(1)},
+		},
+		Slots: []common.SlotUpdate{
+			{Account: addr1, Key: common.Key{0, 0, 0, 1}, Value: value},
+		},
+	}
+
+	state, err := initTestedStates(t)
+	require.NoError(t, err, "failed to initialize tested states")
+	defer func() {
+		require.NoError(t, state.Close(), "failed to close state")
+	}()
+
+	require.NoError(t, state.Apply(0, update))
+
+	hash, err := state.GetHash()
+	require.NoError(t, err)
+	require.NotEmpty(t, hash, "hash should not be empty")
+}
+
+// initTestedStates initializes a comparingState instance
+// for all State instances under test.
+func initTestedStates(t *testing.T) (*comparingState, error) {
+	st := &comparingState{
+		states: make(map[string]state.State),
+		t:      t,
+	}
+
+	gethSt, err1 := geth.NewState(state.Parameters{})
+	refSt, err2 := reference.NewState(state.Parameters{})
+	innerSt, err3 := reference.NewState(state.Parameters{})
+	flatSt, err4 := flat.NewState(t.TempDir(), innerSt)
+
+	st.put("memory", refSt)
+	st.put("geth", gethSt)
+	st.put("flat", flatSt)
```

Comment:
For maintainability, these states should be derived from the variant registry. Everything that declares to be an S6 implementation should be covered.